Feature/hel 20

yixuanG opened the pull request:
## Summary

Integrated team schemas from HEL-18 (Marketing) and HEL-19 (Procurement CSV), and generated a sample of HR data structure for the 3 core departments for an easier preview.

## Key Deliverables

### Team Schema Integration
- **HR**: 3 JSON schemas (employees, skills)
- **Marketing** (from HEL-18): 4 JSON schemas + 3 dictionaries (brands, KPIs, objectives)
- **Procurement** (from HEL-19): 4 JSON schemas + 3 dictionaries (tiers, risks, contracts)

## Data Overview

| Domain      | Nodes | Key Entities                                         |
| ----------- | ----- | ---------------------------------------------------- |
| HR          | 262   | 200 employees, 3 departments, 3 locations, 50 skills |
| Marketing   | 2,456 | 600 campaigns, 1,828 orders, 10 brands               |
| Procurement | 3,446 | 240 suppliers, 120 products, 1,452 POs, 674 invoices |

## How to Run

### Load All Data

```bash
# Start Jupyter notebook
jupyter notebook notebooks/run_etl_local.ipynb

# Then: Kernel → Restart & Run All Cells
```

### Queries & Vizualization
```bash
# Start Jupyter notebook
jupyter notebook notebooks/graph_queries.ipynb

# Then: Kernel → Restart & Run All Cells
```
You will get an HTML file to view the entire 3D graph.

### Validate Schemas

```bash
python scripts/validate_all_schemas.py
```

## Task Checklist

- [X] ETL framework supports all three domains
- [X] 100+ employees with realistic profiles (200)
- [X] 40+ unique skills properly categorized (50)
- [X] All data loaded into Neo4j without errors
- [X] Graph has ≥2,000 nodes total (6,164)
- [X] Indexes and constraints properly configured
- [X] Load time documented (should be <5 minutes) (within 30s)

yixuanG committed:
feat: Add multi-domain ETL framework with HR, Marketing, and Procurement loaders

- Implemented generic BaseLoader class for reusable ETL operations
- Created domain-specific loaders (HRLoader, MarketingLoader, ProcurementLoader)
- Added Pydantic schemas for all three domains with validation
- Generated synthetic datasets: 200 employees, 600 campaigns, 150 suppliers
- Built comprehensive testing suite (schema validation + end-to-end tests)
- Created Jupyter notebooks for ETL execution and graph queries
- Added interactive 3D graph visualization (9,793 nodes, 27,131 relationships)
- Updated documentation and cleaned up redundant files

Total graph: 9,793 nodes, 27,131 relationships across 3 domains

yixuanG committed:
feat: Integrate team schemas and simplify HR data structure

Major Changes:
- Integrated teammate Marketing schemas (HEL-4) and dictionaries
- Integrated teammate Procurement CSV data and schemas (HEL-19)
- Simplified HR data: 6 departments → 3 core departments
- Added ProcurementCSVLoader with DateTime parsing fix

Team Schema Integration:
- Added 8 JSON schemas (4 Marketing + 4 Procurement) in ontologies/
- Added 6 dictionary files in dictionaries/
- Marketing: brands, KPIs, objectives (107 reference items)
- Procurement: supplier tiers, risk categories, contract types

Procurement CSV Integration:
- Replaced JSON data with CSV format (3,446 records)
- Added 5 CSV files: suppliers, products, purchase_orders, invoices, risks
- Created ProcurementCSVLoader with DateTime format handling
- Integrated RiskCalculator business logic
- Added load_procurement_csv.py script

HR Data Simplification:
- Reduced from 6 to 3 core departments: Engineering, Marketing, Sales
- Reduced from 6 to 3 locations: San Francisco, London, Tokyo
- Regenerated 200 employees with new structure
- Updated generate_hr_data.py with expanded job titles

Technical Improvements:
- Fixed DateTime parsing for CSV dates (YYYY-MM-DD HH:MM:SS → ISO 8601)
- Updated etl/__init__.py with ProcurementCSVLoader
- Updated notebooks to use new loaders
- Enhanced .gitignore for better file management
- Added schema validation script

Data Statistics:
- HR: 200 employees, 3 departments, 3 locations
- Marketing: 600 campaigns (kept existing data)
- Procurement: 240 suppliers, 120 products, 1,452 POs, 674 invoices, 960 risks

Documentation:
- Added data_sources_and_schemas.md
- Added TESTING.md
- Added README files for ontologies and dictionaries

Related: HEL-20, HEL-4, HEL-19

datamatta1 reviewed (approved):
Excellent work, all the hard requirements are met. Well architected and prod-quality ETL exceeds many requirements. 

Some of the issues I would like to get fixed before pushed to main: 

- Neo4j: add provenance properties on relationships (created_at, source): source property is missing on all HR relationships

In etl/hr_loader.py, add source property to all relationship creation queries:
```
  SET r1.created_at = datetime(),
      r1.source = "hr_data"     # ADD THIS
```

  Apply to all relationships: WORKS_IN, LOCATED_IN, REPORTS_TO, HAS_SKILL

- Role taxonomy: Engineering, Marketing, Sales, Ops, HR, Finance
Current State:
  - Found only 3 departments: Engineering, Marketing, Sales
  - Missing: Ops, HR, Finance

Update scripts/generate_hr_data.py to include all 6 departments when generating employee data.

- Team assignments (10–15 teams)

  Current State:
  - No evidence of Team nodes in the schema
  - No MEMBER_OF or similar team relationships
  - No team data in JSON files

HR organizational structure is incomplete without teams. 

**Fix Required:**
  1. Add Team entity to schema (schemas/hr_schema.py)
  2. Create Team nodes in data generation
  3. Add MEMBER_OF relationship (Employee → Team)
  4. Generate 10-15 teams with realistic names
  5. Update etl/hr_loader.py to load teams and relationships

After all that is done, feel free to merge the PR to main. Looks good to me.

yixuanG committed:
Fix HR data pipeline issues before merge to main

- Add source property to all HR relationships (WORKS_IN, LOCATED_IN, REPORTS_TO, HAS_SKILL, MEMBER_OF)
- Expand role taxonomy from 3 to 6 departments: add Ops, HR, Finance
- Add Team entity to schema with team_id, name, department, description
- Generate 15 teams across all departments with realistic assignments
- Add MEMBER_OF relationship (Employee → Team) with role property
- Update ETL loader to handle teams and employee-team relationships
- Add team constraints and indexes in Neo4j schema
- Update data generation with complete job titles and salary ranges for new departments

All hard requirements now met:
✓ Neo4j provenance on all relationships
✓ 6-department role taxonomy
✓ 10-15 team structure with assignments
